core/workspace_svg.js

```diff
@@ -1666,6 +1679,30 @@ Blockly.WorkspaceSvg.prototype.getButtonCallback = function(key) {
   return this.flyoutButtonCallbacks_[key];
 };
 
+/**
+ * Register a callback function associated with a given key, for populating
+ * custom toolbox categories in this workspace.  See the variable and procedure
+ * categories as an example.
+ * @param {string} key The name to use to look up this function.
+ * @param {function(!Blockly.Workspace):!Array<!Element>} func The function to
+ *     call when the given toolbox category is opened.
+ */
+Blockly.WorkspaceSvg.prototype.registerToolboxCategoryCallback = function(key,
+    func) {
+  this.toolboxCategoryCallbacks_[key] = func;
+};
+
+/**
+ * Get the callback function associated with a given key, for populating
+ * custom toolbox categories in this workspace.
+ * @param {string} key The name to use to look up the function.
+ * @return {function(!Blockly.Workspace):!Array<!Element>} The function
+ *     corresponding to the given key for this workspace.
+ */
+Blockly.WorkspaceSvg.prototype.getToolboxCategoryCallback = function(key) {
```

Comment:
How do we expect folks to use this? Purely to populate a toolbox category? Would we want to prevent/discourage use to do things like showing a tutorial, animation, or something when a category is opened?  I'm guessing we'd want people to do those kinds of things with the UIEvent instead. Though, I'm not sure there's a way to technically force that. I guess I'm saying that if it is purely for populating a toolbox category with blocks, we could make the name more specific.